Cpp_test/lib/ipc-queue-send.cpp

```diff
@@ -0,0 +1,100 @@
+#include "ipc-queue-send.h"
+#include <ctime>
+#include <iostream>
+
+ipcQueueSend::~ipcQueueSend()
+{
+	mq_close(this->mqd);
+	mq_unlink(this->info.method_name);
+}
+
+void ipcQueueSend::init()
+{
+	// remove old queue name (if any)
+	mq_unlink(this->info.method_name);
+	errno = 0; // clear errno
+	this->mqd = mq_open(this->info.method_name, O_CREAT | O_EXCL | O_WRONLY | O_NONBLOCK, 0660, &(this->attr));
+
+	if (this->mqd == -1)
+		throw std::runtime_error(std::string("ERROR: ") + this->info.method_name + ": " + strerror(errno));
+	else
+		std::cout << this->info.method_name << " is opened." << std::endl;
+
+	this->open_file();
+}
+
+void ipcQueueSend::transfer()
+{
+	std::cout << "transfer\n";
+	long mq_send_return_value{0};
+	long read_bytes{0};
+	std::vector<char> buffer(this->attr.mq_msgsize);
+	unsigned total_sent_bytes{0};
+	unsigned long file_size = this->get_file_size();
+	if (file_size == 0){
+		throw std::runtime_error("ERROR: File size = 0.");
+	}
+		
+	std::cout << "Sending..." << std::endl;
+	while (!this->fs.eof())
+	{
+		this->read_file(buffer, this->attr.mq_msgsize);
+		read_bytes = this->fs.gcount();
+
+		if (read_bytes > 0)
+		{
+			errno = 0; // clear errno
+			mq_send_return_value = mq_send(this->mqd, buffer.data(), read_bytes,
+										   this->priority);
+			if (mq_send_return_value == 0)
+			{
+				total_sent_bytes += read_bytes;
+				
+			}
+			else if (mq_send_return_value == -1 && errno == EAGAIN)
+			{
+				while (errno == EAGAIN )
```

Comment:
This loop will consume a lot of CPU if the program ipc_receive is slow/shutdown or something else. (it's quite like doing a `while(1){}`)

Reply:
I also will add timeout here.